File: tests/refs/palindrome_check.py
"""Reference: palindrome check (case- and punctuation-insensitive).

Matches Week 7's `is_palindrome_ignore_non_alpha` semantics so "A man, a plan,
a canal: Panama" is True.
"""

from __future__ import annotations


def isPalindrome(s: str) -> bool:
    l, r = 0, len(s) - 1
    while l < r:
        while l < r and not s[l].isalnum():
            l += 1
        while l < r and not s[r].isalnum():
            r -= 1
        if s[l].lower() != s[r].lower():
            return False
        l += 1
        r -= 1
    return True

File: tests/refs/PalindromeCheck.java
/*
 * Reference Java implementation for tests/cases/palindrome_check.json.
 * Case- and punctuation-insensitive palindrome check using two pointers.
 */
public class PalindromeCheck {
    public static boolean isPalindrome(String s) {
        int l = 0, r = s.length() - 1;
        while (l < r) {
            while (l < r && !Character.isLetterOrDigit(s.charAt(l))) l++;
            while (l < r && !Character.isLetterOrDigit(s.charAt(r))) r--;
            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))) {
                return false;
            }
            l++; r--;
        }
        return true;
    }
}
